### PR TITLE
bluestore/BitAllocator: miscellaneous fixes

### PR DESCRIPTION
**Pull request by xiexingguo**

### CODE REVIEW

**Comment by xiexingguo** [comment]
retest this please

**Comment by xiexingguo** [comment]
@chhabaramesh Ramesh, more fixes and cleanups about the newly merged Bitmap-Allocator. Could you take a look again? Thanks!

**Comment by chhabaramesh** [comment]
@xiexingguo I will review today.

File: src/os/bluestore/BitAllocator.cc [hunk 1]
```diff
@@ -1596,7 +1596,7 @@ void BitAllocator::free_blocks_dis(int64_t num_blocks, int64_t *block_list)
     free_blocks_int(block_list[i], 1);
   }
 
-  debug_assert(get_used_blocks() > 0);
   sub_used_blocks(num_blocks);
+  debug_assert(get_used_blocks() >= 0);
```
Comment (chhabaramesh) [inline src/os/bluestore/BitAllocator.cc]: looks good

File: src/os/bluestore/BitAllocator.cc [hunk 2]
```diff
@@ -600,8 +600,8 @@ int64_t BitMapZone::alloc_blocks(int64_t num_blocks, int64_t *start_block)
 void BitMapZone::free_blocks(int64_t start_block, int64_t num_blocks)
 {
   free_blocks_int(start_block, num_blocks);
-  debug_assert(get_used_blocks() > 0);
   sub_used_blocks(num_blocks);
+  debug_assert(get_used_blocks() >= 0);
```
Comment (chhabaramesh) [inline src/os/bluestore/BitAllocator.cc]: looks good

File: src/os/bluestore/BitAllocator.cc [hunk 3]
```diff
@@ -1547,13 +1549,13 @@ int64_t BitAllocator::alloc_blocks_dis(int64_t num_blocks, int64_t *block_list)
...
+  if (is_stats_on()) {
+    m_stats->add_concurrent_scans(scans);
+  }
 
   while (scans && allocated < num_blocks) {
     allocated += alloc_blocks_dis_int(false, num_blocks, blk_off, &block_list[allocated]);
-    scans --;
-  }
-  if (is_stats_on()) {
-    m_stats->add_concurrent_scans(scans);
```
Comment (chhabaramesh) [inline src/os/bluestore/BitAllocator.cc]: This commit look good.

File: src/os/bluestore/BitAllocator.cc [hunk 4]
```diff
@@ -826,6 +826,7 @@ void BitMapAreaIN::unreserve(int64_t needed, int64_t allocated)
 }
 int64_t BitMapAreaIN::get_reserved_blocks()
 {
+  std::lock_guard<std::mutex> l(m_blocks_lock);
```
Comment (chhabaramesh) [inline src/os/bluestore/BitAllocator.cc]: Looks good

File: src/os/bluestore/BitAllocator.cc [hunk 5]
```diff
@@ -843,8 +843,6 @@ bool BitMapAreaIN::is_allocated(int64_t start_block, int64_t num_blocks)
     return true;
   }
 
-  assert(start_block >= 0);
```
Comment (chhabaramesh) [inline src/os/bluestore/BitAllocator.cc]: this commit looks good

File: src/os/bluestore/BitAllocator.cc [hunk 6]
```diff
@@ -971,7 +971,7 @@ void BitMapAreaIN::set_blocks_used_int(int64_t start_block, int64_t num_blocks)
   int64_t blks = num_blocks;
   int64_t start_blk = start_block;
 
-  assert(start_block >= 0);
+  debug_assert(start_block >= 0);
```
Comment (chhabaramesh) [inline src/os/bluestore/BitAllocator.cc]: Looks good

File: src/os/bluestore/BitAllocator.cc [hunk 7]
```diff
@@ -703,11 +703,8 @@ void BitMapAreaIN::init(int64_t total_blocks, int64_t area_idx, bool def)
   init_common(total_blocks, area_idx, def);
   int64_t level_factor = pow(BitMapArea::get_span_size(), m_level);
 
-  num_child = total_blocks / level_factor;
+  num_child = (total_blocks + level_factor - 1) / level_factor;
   debug_assert(num_child < MAX_INT16);
-  if (total_blocks % level_factor) {
```
Comment (chhabaramesh) [inline src/os/bluestore/BitAllocator.cc]: This is not needed. Both are same :).

File: src/os/bluestore/BitAllocator.cc [hunk 8]
```diff
@@ -617,9 +617,6 @@ int64_t BitMapZone::alloc_blocks_dis(int64_t num_blocks, int64_t zone_blk_off, i
 
   debug_assert(check_locked());
 
-  bmap_idx = 0;
-  bit = 0;
```
Comment (chhabaramesh) [inline src/os/bluestore/BitAllocator.cc]: looks good

File: src/os/bluestore/BitAllocator.cc [hunk 9]
```diff
@@ -92,7 +92,7 @@ int64_t BmapEntityListIter::index()
 void BmapEntityListIter::decr_idx()
 {
   m_cur_idx--;
-  debug_assert(m_cur_idx > 0);
+  debug_assert(m_cur_idx >= 0);
```
Comment (chhabaramesh) [inline src/os/bluestore/BitAllocator.cc]: looks good

File: src/os/bluestore/BitAllocator.h [hunk 10]
```diff
@@ -20,7 +20,6 @@
 
 class BitAllocatorStats {
 public:
-  bool m_on;
```
Comment (chhabaramesh) [inline src/os/bluestore/BitAllocator.h]: looks good

File: src/os/bluestore/BitAllocator.h [hunk 11]
```diff
@@ -126,7 +126,7 @@ class BitMapEntityIter {
   }
   void decr_idx() {
     m_cur_idx--;
-    debug_assert(m_cur_idx > 0);
+    debug_assert(m_cur_idx >= 0);
```
Comment (chhabaramesh) [inline src/os/bluestore/BitAllocator.h]: Looks good

File: src/os/bluestore/BitAllocator.cc [hunk 12]
```diff
@@ -5,15 +5,17 @@
...
  *
  * Rest of the nodes are classified in to three catagories:
- *   root note or Alloctor, internal nodes or BitMapAreaIN and
- *  finally nodes that contains Zones called BitMapAreaLeaf.
- * This classification is according some their own implmentation of some the interfaces define in
- * BitMapArea.
+ *   root note or Alloctor
+ *   internal nodes or BitMapAreaIN
+ *   finally nodes that contains Zones called BitMapAreaLeaf
+ * This classification is according to some their own implmentation of
```
Comment (chhabaramesh) [inline src/os/bluestore/BitAllocator.cc]: remove some before their

File: src/os/bluestore/BitAllocator.cc [hunk 13]
```diff
@@ -71,7 +73,7 @@ BitMapArea* BmapEntityListIter::next()
 
   if (m_cur_idx == m_list->size() &&
       m_wrap) {
-    m_cur_idx %= m_list->size();
```
Comment (chhabaramesh) [inline src/os/bluestore/BitAllocator.cc]: Look ok

File: src/os/bluestore/BitAllocator.h [hunk 14]
```diff
@@ -97,7 +97,6 @@ class BitMapEntityIter {
        */
       if (!m_end) {
         m_end = true;
-        //return m_list->get_nth_item(cur_idx);
```
Comment (chhabaramesh) [inline src/os/bluestore/BitAllocator.h]: Looks ok

File: src/os/bluestore/BitAllocator.h [hunk 15]
```diff
@@ -106,7 +105,7 @@ class BitMapEntityIter {
 
     if (m_cur_idx == m_list->size() &&
         m_wrap) {
-      m_cur_idx %= m_list->size();
+      m_cur_idx = 0;
```
Comment (chhabaramesh) [inline src/os/bluestore/BitAllocator.h]: Looks ok

File: src/os/bluestore/BitAllocator.cc [hunk 16]
```diff
@@ -5,15 +5,17 @@
...
- * Storage is divided into bitmap of blocks. Each bitmap has size of unsigned long.
- * Group of bitmap creates a Zone. Zone is a unit where at a time single
- * thread can be active as well as single biggest contiguous allocation that can be requested.
+ * Storage is divided into bitmap of blocks. Each bitmap has size of
+ * unsigned long. Group of bitmap creates a Zone. Zone is a unit where
+ * at a time single thread can be active as well as single biggest
+ * contiguous allocation that can be requested.
  *
  * Rest of the nodes are classified in to three catagories:
- *   root note or Alloctor, internal nodes or BitMapAreaIN and
```
Comment (chhabaramesh) [inline src/os/bluestore/BitAllocator.cc]: Alloctor -> Allocator

**Comment by chhabaramesh** [comment]
@xiexingguo  Overall changes looks good.  @liewegas  changes ok from my side.

**Comment by xiexingguo** [comment]
@chhabaramesh Fixed the typo. Thanks for the review.

File: src/os/bluestore/BitAllocator.cc [hunk 17]
```diff
@@ -11,11 +11,11 @@
  * contiguous allocation that can be requested.
  *
  * Rest of the nodes are classified in to three catagories:
- *   root note or Alloctor
+ *   root note or Allocator
  *   internal nodes or BitMapAreaIN
  *   finally nodes that contains Zones called BitMapAreaLeaf
- * This classification is according to some their own implmentation of
- * some the interfaces define in BitMapArea.
+ * This classification is according to their own implmentation of some
```
Comment (chhabaramesh) [inline src/os/bluestore/BitAllocator.cc]: Looks good

**Comment by chhabaramesh** [comment]
@xiexingguo  Looks good. Should it be squashed in to single change? 
up to you and @liewegas  to decide.

**Comment by xiexingguo** [comment]
Squashed. Sorry for the laziness and thanks!

**Comment by xiexingguo** [comment]
@liewegas This pr is safe to merge now.

**Pull request merged**